src/storage/v3/shard_rsm.cpp

```diff
@@ -109,6 +111,299 @@ Value ConstructValueVertex(const memgraph::storage::v3::VertexAccessor &acc, mem
+                      req.transaction_id.logical_id);
+        return std::nullopt;
+      }
+      out_edges = out_edges_result.GetValue();
+      break;
+    }
+  }
+  return std::array<std::vector<memgraph::storage::v3::EdgeAccessor>, 2>{in_edges, out_edges};
+}
+
+using AllEdgePropertyDataSructure = std::map<PropertyId, memgraph::msgs::Value>;
+using SpecificEdgePropertyDataSructure = std::vector<memgraph::msgs::Value>;
+
+using AllEdgeProperties = std::tuple<memgraph::msgs::VertexId, memgraph::msgs::Gid, AllEdgePropertyDataSructure>;
+using SpecificEdgeProperties =
+    std::tuple<memgraph::msgs::VertexId, memgraph::msgs::Gid, SpecificEdgePropertyDataSructure>;
+
+using SpecificEdgePropertiesVector = std::vector<SpecificEdgeProperties>;
+using AllEdgePropertiesVector = std::vector<AllEdgeProperties>;
+
+template <typename ReturnType, typename EdgeProperties, typename EdgePropertyDataStructure, typename Functor>
+std::optional<ReturnType> GetEdgesWithProperties(const std::vector<memgraph::storage::v3::EdgeAccessor> &edges,
+                                                 const memgraph::msgs::ExpandOneRequest &req,
+                                                 Functor get_edge_properties) {
+  ReturnType ret;
+  ret.reserve(edges.size());
+
+  for (const auto &edge : edges) {
+    if (!DoesEdgeTypeMatch(req, edge)) {
+      continue;
+    }
+
+    EdgeProperties ret_tuple;
+
+    memgraph::msgs::Label label;
+    label.id = edge.FromVertex().primary_label;
+    memgraph::msgs::VertexId other_vertex = std::make_pair(label, ConvertValueVector(edge.FromVertex().primary_key));
+
+    const auto edge_props_var = get_edge_properties(edge);
+
+    if (std::get_if<LocalError>(&edge_props_var) != nullptr) {
+      return std::nullopt;
+    }
+
+    auto edge_props = std::get<EdgePropertyDataStructure>(edge_props_var);
+    memgraph::msgs::Gid gid = edge.Gid().AsUint();
+
+    ret_tuple = {other_vertex, gid, edge_props};
+    ret.push_back(ret_tuple);
+  }
+
+  return ret;
+}
+
+template <typename PropertyValueT, typename PropertyNulloptT>
+void SetFinalEdgeProperties(std::optional<PropertyValueT> &properties_to_value,
+                            std::optional<PropertyNulloptT> &properties_to_nullopt, const PropertyValueT &ret_out,
+                            const PropertyValueT &ret_in, const memgraph::msgs::ExpandOneRequest &req) {
+  switch (req.direction) {
+    case memgraph::msgs::EdgeDirection::OUT: {
+      properties_to_value = ret_out;
+      break;
+    }
+    case memgraph::msgs::EdgeDirection::IN: {
+      properties_to_value = ret_in;
+      break;
+    }
+    case memgraph::msgs::EdgeDirection::BOTH: {
+      PropertyValueT ret;
+      ret.resize(ret_out.size() + ret_in.size());
+      ret.insert(ret.end(), ret_in.begin(), ret_in.end());
+      ret.insert(ret.end(), ret_out.begin(), ret_out.end());
+
+      properties_to_value = ret;
+      break;
```

Comment:
What about moving/forwarding? I can assume copying these datastructures might be expensive. Same true for other places in the new code.